engine/crates/engine-v2/src/executor/graphql/query.rs

```diff
@@ -0,0 +1,152 @@
+            OperationType::Subscription => write!(out, "subscription ")?,
+        };
+
+        let operation_name = "Subgraph"; // TODO: should re-use the real operation name
+        out.push_str(operation_name);
+        out.push_str(&buffer.inner);
+
+        Ok(out)
+    }
+
+    fn write_selection_set(&self, buffer: &mut Buffer, selection_set: &OperationSelectionSet) -> Result<(), Error> {
+        if !selection_set.is_empty() {
+            buffer.write(" {\n")?;
+            buffer.indent += 1;
+            let gouped_by_type_condition = selection_set
+                .iter()
+                .map(|selection| (&self.fields[selection.operation_field_id].type_condition, selection))
+                .group_by(|(type_condition, _)| *type_condition);
+            for (type_condition, selections) in &gouped_by_type_condition {
+                let selections = selections.into_iter().map(|(_, selection)| selection);
+                if let Some(type_condition) = type_condition {
+                    self.write_inline_fragment(buffer, *type_condition, selections)?;
+                } else {
+                    for selection in selections {
+                        self.write_selection(buffer, selection)?;
+                    }
+                }
+            }
+            buffer.indent -= 1;
+            buffer.indent_write("}\n")?;
+        }
+        Ok(())
+    }
+
+    fn write_inline_fragment<'b>(
+        &self,
+        buffer: &mut Buffer,
+        type_condition: TypeCondition,
+        selections: impl IntoIterator<Item = &'b OperationSelection>,
+    ) -> Result<(), Error> {
+        buffer.indent_write("... on ")?;
+        buffer.write(match type_condition {
+            TypeCondition::Interface(interface_id) => self.names.interface(interface_id),
+            TypeCondition::Object(object_id) => self.names.object(object_id),
+            TypeCondition::Union(union_id) => self.names.union(union_id),
+        })?;
+        buffer.write(" {\n")?;
+        buffer.indent += 1;
+        for selection in selections {
+            self.write_selection(buffer, selection)?;
+        }
+        buffer.indent -= 1;
+        buffer.indent_write("}\n")?;
+        Ok(())
+    }
+
+    fn write_selection(&self, buffer: &mut Buffer, selection: &OperationSelection) -> Result<(), Error> {
+        let field = &self.fields[selection.operation_field_id];
+        let name = self.names.field(field.field_id);
+        buffer.indent_write(name)?;
+        self.write_selection_set(buffer, &selection.subselection)?;
+        Ok(())
+    }
+
+    fn write_arguments(&self, buffer: &mut Buffer, arguments: &Vec<OperationArgument>) -> Result<(), Error> {
+        if !arguments.is_empty() {
+            buffer.write("(")?;
+
+            let mut arguments = arguments.iter().peekable();
+            while let Some(argument) = arguments.next() {
+                buffer.add_variable_references(argument.value.variables_used().map(|name| name.to_string()));
```

Comment:
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.format_with could make this a bit nicer (and reallocation-efficient!), but not a huge issue

Reply:
Same in the other functions.

Reply:
Oh interesting, will take a look at it when extend this with directives & fragments. thanks!